### PR TITLE
Vulkan Renderer - implement YcBcCr using VK_KHR_sampler_ycbcr_conversion.

### PR DESCRIPTION
**Pull request by danginsburg**
## Description

This simplifies the shader code and will also allow support for additional formats that we don't yet support (such as SDL_PIXELFORMAT_P010 for ffmpeg). The renderer now queries for VK_KHR_sampler_ycbcr_conversion and dependent extensions and will enable it if it's present. It reimplements YUV/NV12 texture support using this extension (these formats are no longer supported if the extension is not present). For each YUV/NV12 textures, a VkSamplerYcbcrConversion object is created from SDL_Colorspace parameters.  This is passed to the VkImageView and also an additional sampler is created for Ycbcr.  Instead of using 1-3 textures, the shaders now all use 1 texture with a combined image sampler (required by the extension).  Further, when using Ycbcr, a separate descriptor set layout is baked with the Ycbcr sampler as an immutable sampler.  The code to copy the images now copies to the individual image planes.  The swizzling between formats is handled in the VkSamplerYcbcrConversion object.

### CODE REVIEW

File: src/render/vulkan/SDL_render_vulkan.c [hunk 1]
```diff
@@ -1574,14 +1672,18 @@ static VkResult VULKAN_CreateDeviceResources(SDL_Renderer *renderer, SDL_Propert
...
+        const char **instanceExtensionsCopy = SDL_calloc(instanceCreateInfo.enabledExtensionCount + 2, sizeof(const char *));
         for (uint32_t i = 0; i < instanceCreateInfo.enabledExtensionCount; i++) {
             instanceExtensionsCopy[i] = instanceExtensions[i];
         }
         if (rendererData->supportsEXTSwapchainColorspace) {
             instanceExtensionsCopy[instanceCreateInfo.enabledExtensionCount] = VK_EXT_SWAPCHAIN_COLOR_SPACE_EXTENSION_NAME;
             instanceCreateInfo.enabledExtensionCount++;
         }
+        if (rendererData->supportsKHRGetPhysicalDeviceProperties2) {
+            instanceExtensionsCopy[instanceCreateInfo.enabledExtensionCount] = SDL_strdup(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
```
Comment (slouken) [inline src/render/vulkan/SDL_render_vulkan.c:1684]: This should not be strdup()

File: src/render/vulkan/SDL_render_vulkan.c [hunk 2]
```diff
@@ -1642,13 +1746,21 @@ static VkResult VULKAN_CreateDeviceResources(SDL_Renderer *renderer, SDL_Propert
...
+             /* VK_KHR_sampler_ycbcr_conversion + dependent extensions */
+            VK_KHR_SAMPLER_YCBCR_CONVERSION_EXTENSION_NAME,
+            VK_KHR_MAINTENANCE1_EXTENSION_NAME,
+            VK_KHR_BIND_MEMORY_2_EXTENSION_NAME,
+            VK_KHR_GET_MEMORY_REQUIREMENTS_2_EXTENSION_NAME,
         };
+        /* Check if YcBcr extensions are supported */
+        rendererData->supportsKHRSamplerYcBcrConversion = rendererData->supportsKHRGetPhysicalDeviceProperties2 &&
 
+        VULKAN_DeviceExtensionsFound(rendererData, 4, &deviceExtensionNames[1]);
```
Comment (slouken) [inline src/render/vulkan/SDL_render_vulkan.c:1758]: Can you change this to if (...)? It's not immediately clear what's happening, especially given the indentation.

File: src/render/vulkan/SDL_render_vulkan.c [hunk 3]
```diff
@@ -2384,13 +2437,119 @@ static int VULKAN_CreateTexture(SDL_Renderer *renderer, SDL_Texture *texture, SD
...
+            break;
+        }
+
+        switch (SDL_COLORSPACECHROMA(texture->colorspace)) {
+        case SDL_CHROMA_LOCATION_LEFT:
+            samplerYcbcrConversionCreateInfo.xChromaOffset = VK_CHROMA_LOCATION_COSITED_EVEN_KHR;
+            samplerYcbcrConversionCreateInfo.yChromaOffset = VK_CHROMA_LOCATION_MIDPOINT_KHR;
+            break;
+        case SDL_CHROMA_LOCATION_NONE:
+        case SDL_CHROMA_LOCATION_CENTER:
```
Comment (slouken) [inline src/render/vulkan/SDL_render_vulkan.c:2501]: default: ?

File: src/render/vulkan/VULKAN_PixelShader_Common.incl [hunk 4]
```diff
@@ -11,7 +8,7 @@ struct PixelShaderInput
     float4 color : COLOR0;
 };
 
-// These should mirror the definitions in SDL_render_d3d12.c
+// These should mirror the definitions in VULKAN_render_d3d12.c
```
Comment (slouken) [inline src/render/vulkan/VULKAN_PixelShader_Common.incl:11]: I'm pretty sure this comment is wrong. ;)

File: src/render/vulkan/VULKAN_PixelShader_Common.incl [hunk 5]
```diff
@@ -120,45 +117,8 @@ float4 GetInputColor(PixelShaderInput input)
...
-        rgba.g = dot(yuv, Gcoeff.xyz);
-        rgba.b = dot(yuv, Bcoeff.xyz);
-        rgba.a = 1.0;
     } else {
-        // Error!
-        rgba.r = 1.0;
-        rgba.g = 0.0;
-        rgba.b = 0.0;
-        rgba.a = 1.0;
+        rgba = texture0.Sample(sampler0, input.tex).rgba;
```
Comment (slouken) [inline src/render/vulkan/VULKAN_PixelShader_Common.incl:121]: The texture type constants can be removed from here and SDL_render_vulkan.c, right?

File: src/render/vulkan/SDL_render_vulkan.c (hunk 1 above)
Comment (danginsburg) [reply src/render/vulkan/SDL_render_vulkan.c]: Fixed.

File: src/render/vulkan/SDL_render_vulkan.c (hunk 3 above)
Comment (danginsburg) [reply src/render/vulkan/SDL_render_vulkan.c:2502]: Fixed

File: src/render/vulkan/VULKAN_PixelShader_Common.incl (hunk 5 above)
Comment (danginsburg) [reply src/render/vulkan/VULKAN_PixelShader_Common.incl]: Fixed

**Pull request merged**